### PR TITLE
Clang Format file and script

### PR DESCRIPTION
**Pull request by pmatos**
Adds paths to ignore to clang-format, .clang-format file. Also a wrapper script to clang-format to read .clang-format-ignore.

To format the whole tree at the root of the repository run: `find . -iname '*.h' -o -iname '*.cpp' -exec python3 Scripts/clang-format.py -i \{\} \;`

### CODE REVIEW

**Comment by pmatos** [comment]
**why clang-format.py?**

for older versions of clang-format there was no way to ignore certain files in the tree. clang-format 19 adds this feature: https://clang.llvm.org/docs/ClangFormat.html#clang-format-ignore

Once we can all move to clang-format 19, we can drop the script which emulates this feature in python.

File: CMakeLists.txt
Comment (neobrain) [inline CMakeLists.txt:402]: Could you limit this to our actual source directories?

I'd prefer not having this command make assumptions about what name the build folder has (I actually use multiple) or what other files are in the source folder.

Instead of `find`, maybe `git ls-tree` would be a better tool to use here?

Comment (neobrain) [reply CMakeLists.txt:402]: Ah, `git ls-files` is an even better match, since it has a `-c` toggle with which the index is included as well.

Comment (pmatos) [reply CMakeLists.txt:402]: yes - git ls-files is definitely better, but why is the difference so large between what you reported on discord and my reformatting? Let me look at it.

File: Scripts/clang-format.py
Comment (neobrain) [inline Scripts/clang-format.py:66]: Adding `--verbose` here might be useful to get some feedback (1 line per file). Maybe my VM is IO bottlenecked, but reformatting the entire tree takes ~15s without printing anything otherwise.

Comment (neobrain) [reply Scripts/clang-format.py:66]: Actually realized you can just pass this in manually since argv is passed along from the script to `clang-format`. Let's leave it as is :)

File: CMakeLists.txt
Comment (pmatos) [reply CMakeLists.txt]: I have updated this now. I was a bit worried about putting a bash script in the CMakeLists command so I moved that to a `reformat.sh` script. What do you think?

File: CMakeLists.txt
Comment (Sonicadvance1) [inline CMakeLists.txt:350]: This needs to be changed to
```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 6f5e9081a..49c1a2972 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -347,7 +347,7 @@ if (ENABLE_IWYU)
 endif()

 add_custom_target(reformat
-    COMMAND Scripts/reformat.sh ${CMAKE_SOURCE_DIR}
+  COMMAND ${CMAKE_CURRENT_SOURCE_DIR}/Scripts/reformat.sh ${CMAKE_SOURCE_DIR}
 )

 add_compile_options(-Wall)
```
Otherwise when executing from a build folder it doesn't find the script

File: Scripts/reformat.sh
Comment (Sonicadvance1) [inline Scripts/reformat.sh:15]: Changing this to
```diff
diff --git a/Scripts/reformat.sh b/Scripts/reformat.sh
index 957d7256c..7c37f3a6e 100755
--- a/Scripts/reformat.sh
+++ b/Scripts/reformat.sh
@@ -10,9 +10,6 @@ fi

 # Reformat whole tree.
 # This is run by the reformat target.
-for file in `git ls-files '*.cpp' '*.h'`
-do
-    python3 Scripts/clang-format.py -i ${file}
-done
+git ls-files -z '*.cpp' '*.h' | xargs -0 -n 1 -P $(nproc) python3 Scripts/clang-format.py -i

-cd $DIR
\ No newline at end of file
+cd $DIR
```
Allows the script to be run threaded which significantly improves execution time.

Comment (pmatos) [reply Scripts/reformat.sh]: You're right - done.

File: CMakeLists.txt
Comment (pmatos) [reply CMakeLists.txt]: Thanks.

**Review by Sonicadvance1 — APPROVED** [review]

**Comment by pmatos** [comment]
@Sonicadvance1 landing the CI without this PR will format incorrectly because this PR is the one with the `.clang-format` file for the project. 

Sorry for not mentioning this PR dependency. :(

**Comment by Sonicadvance1** [comment]
Nice, no worries. I don't see anything blocking this aside from final pending review.

**Review by neobrain — APPROVED** [review]
Woops, still had one pending comment left I thought I'd already submitted.

Does the CMake target serve much of a purpose now that there's a shell script? It's longer to type `ninja reformat` and you can't actually run it without a build folder, so it doesn't really seem to add any value.

Anyway, looks fine otherwise. Tested out the command again and the unexpected changes I reported on Discord are now fully gone since the fixup PR #3577.

**Review by neobrain — CHANGES_REQUESTED** [review]
(changing review status to avoid confusion)

**Comment by pmatos** [comment]
> Woops, still had one pending comment left I thought I'd already submitted.
> 
> Does the CMake target serve much of a purpose now that there's a shell script? It's longer to type `cmake reformat` and you can't actually run it without a build folder, so it doesn't really seem to add any value.
> 
> Anyway, looks fine otherwise. Tested out the command again and the unexpected changes I reported on Discord are now fully gone since the fixup PR #3577. 
> 

It was @Sonicadvance1 who wanted a ninja target for reformatting. Will leave it up to him.

**Comment by Sonicadvance1** [comment]
I would still prefer the cmake target but other people have stronger opinions than me so if they are adamant about it then just delete it.

**Comment by pmatos** [comment]
> I would still prefer the cmake target but other people have stronger opinions than me so if they are adamant about it then just delete it.

Personally, I wouldn't use either so I don't care. @neobrain ?

File: Scripts/clang-format.py
Comment (pmatos) [reply Scripts/clang-format.py:4]: Done.

**Comment by pmatos** [comment]
Ignore the check code formatting fail until this lands. :) It's the one that contains the one true .clang-format file.

**Comment by neobrain** [comment]
> I would still prefer the cmake target but other people have stronger opinions than me so if they are adamant about it then just delete it.
> 
> Personally, I wouldn't use either so I don't care. @neobrain ?

Great, let's proceed with deleting it then. The CMakeLists.txt is already bloated enough as is :)

**Comment by pmatos** [comment]
> > I would still prefer the cmake target but other people have stronger opinions than me so if they are adamant about it then just delete it.
> > Personally, I wouldn't use either so I don't care. @neobrain ?
> 
> Great, let's proceed with deleting it then. The CMakeLists.txt is already bloated enough as is :)

Done. Happy to see this landing soon.

**Review by neobrain — APPROVED** [review]

**Pull request merged**
